crates/hyperdrive-math/src/lp.rs

```diff
@@ -4,9 +4,144 @@ use ethers::types::{I256, U256};
 use fixed_point::FixedPoint;
 use fixed_point_macros::{fixed, int256};
 
-use crate::{State, YieldSpace};
+use crate::{calculate_effective_share_reserves, State, YieldSpace};
 
 impl State {
+    // Calculates the lp_shares for a given contribution when adding liquidity.
+    pub fn calculate_add_liquidity(
+        &self,
+        current_block_timestamp: U256,
+        contribution: FixedPoint,
+        min_apr: FixedPoint,
+        max_apr: FixedPoint,
+        as_base: bool,
+    ) -> Result<FixedPoint, &str> {
+        // Ensure that the contribution is greater than or equal to the minimum
+        // transaction amount.
+        if contribution < self.minimum_transaction_amount() {
+            return Err("MinimumTransactionAmount");
+        }
+
+        // Enforce the slippage guard.
+        let apr = self.calculate_spot_rate();
+        if apr < min_apr || apr > max_apr {
+            return Err("InvalidApr");
```

Comment:
I think something like "Pool APR is outside the provided constraints." would be more clear to the user.

Reply:
Ahh is `InvalidApr` the solidity error? I don't think we have been matching their error codes exactly in any of the other rust implementations, although that is an interesting idea... Is that your intent here? I'm wondering if we could do both? Like include the SOL error code but also add context.

Reply:
Yeah that was the intent, makes matching errors easier when debugging.  Happy to do what you propose and have it be more clear like "InvalidApr: apr is outside the provided constraints."